### PR TITLE
feat: source report entity and unified sending of reports

### PR DESCRIPTION
**Pull request by sshanzel**
- Introduced Source Report Entity
- Consolidated the types for reporting
- Send report mutation by supporting two types

MI-496 #done

### CODE REVIEW

File: src/common/reporting.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
+import { DataSource } from 'typeorm';
```
Comment (sshanzel) [inline src/common/reporting.ts:1]: Everything except `reportSource` was purely taken from `schema/posts`.

File: src/entity/PostReport.ts [hunk 2]
```diff
@@ -14,8 +15,8 @@ export class PostReport {
   @Column({ default: () => 'now()' })
   createdAt: Date;
 
-  @Column({ length: 12 })
-  reason: string;
+  @Column({ length: 12, type: 'varchar' })
+  reason: ReportReason;
```
Comment (sshanzel) [inline src/entity/PostReport.ts:19]: Using an `enum` a column, postgresql thinks it is a `text` type. But when you set a max length of the column, it is automatically turned into `varchar`. Using an `enum` with `length` makes the build go crazy. Specifying the type fixes it.

File: src/schema/posts.ts [hunk 3]
```diff
@@ -1302,40 +1231,8 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
-
-      if (reason === 'IRRELEVANT' && !tags?.length) {
-        throw new ValidationError('You must include the irrelevant tags!');
-      }
-
-      const added = await saveHiddenPost(ctx.con, {
-        userId: ctx.userId,
-        postId: id,
-      });
+      await reportPost({ ctx, id, reason, comment, tags });
```
Comment (sshanzel) [inline src/schema/posts.ts:1234]: To keep backwards compatibility, we are keeping this mutation but using the new function now.

File: src/common/reporting.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  ctx,
+  id,
+  reason,
+  comment,
+}: BaseReportArgs) => {
+  if (!sourceReportReasonsMap.has(reason)) {
+    throw new ValidationError('Reason is invalid');
+  }
+
+  await ctx.getRepository(Source).findOneByOrFail({ id });
```
Comment (capJavert) [inline src/common/reporting.ts:114]: why empty call to this? I think `ensureSourcePermissions` already queries source under the hood.

File: src/entity/sources/SourceReport.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  sourceId: string;
+
+  @PrimaryColumn({ length: 36 })
+  @Index('IDX_source_report_user_id')
+  userId: string;
+
+  @Column({ default: () => 'now()' })
+  createdAt: Date;
+
+  @Column({ length: 32, type: 'varchar' })
```
Comment (capJavert) [inline src/entity/sources/SourceReport.ts:18]: I think we make it text lately since varchar and text are basically the same in psql.

Comment (capJavert) [reply src/entity/sources/SourceReport.ts:18]: same for other columns

File: src/entity/PostReport.ts (hunk 2 above)
Comment (capJavert) [reply src/entity/PostReport.ts:19]: 👀

File: src/entity/sources/SourceReport.ts (hunk 5 above)
Comment (sshanzel) [reply src/entity/sources/SourceReport.ts:18]: Sure, let's update these ones to that.

Comment (sshanzel) [reply src/entity/sources/SourceReport.ts]: Updated this one, but for the others, I had to keep it, the migration says it needs to drop the column first to change the type.

File: src/entity/common.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+export enum ReportEntity {
```
Comment (rebelchris) [inline src/entity/common.ts:1]: Also here, we not fixing comment reports?

File: src/entity/sources/SourceReport.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import { Column, Entity, Index, ManyToOne, PrimaryColumn } from 'typeorm';
+import type { User } from '../user';
+import { ReportReason } from '../common';
+
+@Entity()
```
Comment (rebelchris) [inline src/entity/sources/SourceReport.ts:5]: Do we want to go seperate route, or stick with childEntities?
(Probably was a discussion I can't recall)

File: src/schema/users.ts [hunk 8]
```diff
@@ -925,6 +939,32 @@ export const typeDefs = /* GraphQL */ `
       vote: Int!
     ): EmptyResponse @auth
 
+    """
+    When a user tries to report an entity. The entity can be either post/comment/source or user when we get there
```
Comment (rebelchris) [inline src/schema/users.ts:943]: Here it mentions comment though

**Review by rebelchris — CHANGES_REQUESTED** [review]
Mainly missing comments, but maybe I missed some coms

File: src/entity/common.ts (hunk 6 above)
Comment (sshanzel) [reply src/entity/common.ts:1]: Should be updated now.

File: src/entity/sources/SourceReport.ts (hunk 7 above)
Comment (sshanzel) [reply src/entity/sources/SourceReport.ts:5]: Was near the end of the RFC, but yeah, it was decided to go similar to vote entities.

**Review by rebelchris — APPROVED** [review]
Looks good, just note Ante's comment

File: src/common/reporting.ts (hunk 4 above)
Comment (sshanzel) [reply src/common/reporting.ts:149]: Right. This is just redundant.

**Review by nensidosari — APPROVED** [review]

**Comment by sshanzel** [comment]
@capJavert comments should've been addressed by now 🙏

**Comment by sshanzel** [comment]
@capJavert since the comments have now been addressed, I will merge the PR now so we can move forward with this one 🙏

**Pull request merged**